File: src/game/inventory/destroy.ts
import type { PlayerSave } from '../save/types'

/** Remove bag stacks at the given inventory indexes. Equipped items are untouched. */
export function destroyInventoryIndexes(save: PlayerSave, indexes: Iterable<number>): PlayerSave {
  const remove = new Set(
    [...indexes].filter((index) => Number.isInteger(index) && index >= 0 && index < save.inventory.length),
  )
  if (remove.size === 0) return save
  return {
    ...save,
    inventory: save.inventory.filter((_, index) => !remove.has(index)),
  }
}

File: packages/ik_rules/lib/src/inventory/destroy.dart
import '../js_compat.dart';
import '../save/generated/save_models.dart';

/// Removes bag stacks at the given inventory indexes. Equipped items are untouched.
PlayerSave destroyInventoryIndexes(PlayerSave save, Iterable<num> indexes) {
  final remove = <int>{};
  for (final index in indexes) {
    if (jsIsInteger(index) && index >= 0 && index < save.inventory.length) {
      remove.add(index.toInt());
    }
  }
  if (remove.isEmpty) return save;
  return save.copyWith(
    inventory: save.inventory.indexed
        .where((entry) => !remove.contains(entry.$1))
        .map((entry) => entry.$2)
        .toList(),
  );
}
